presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/PruneExchangeColumns.java

```diff
@@ -0,0 +1,120 @@
+ *        partitioning by (o2)
+ *        hash h
+ *        inputs [[a1, a2, h1], [b1, b2, h2]]
+ *          - source [a1, a2, a3, h1]
+ *          - source [b1, b2, b3, h2]
+ * </pre>
+ */
+public class PruneExchangeColumns
+        extends ProjectOffPushDownRule<ExchangeNode>
+{
+    public PruneExchangeColumns()
+    {
+        super(exchange());
+    }
+
+    @Override
+    protected Optional<PlanNode> pushDownProjectOff(Context context, ExchangeNode exchangeNode, Set<Symbol> referencedOutputs)
+    {
+        // Extract output symbols referenced by parent node or used for partitioning, ordering or as a hash symbol of the Exchange
+        ImmutableSet.Builder<Symbol> builder = ImmutableSet.builder();
+        builder.addAll(referencedOutputs);
+        builder.addAll(exchangeNode.getPartitioningScheme().getPartitioning().getColumns());
+        exchangeNode.getPartitioningScheme().getHashColumn().ifPresent(builder::add);
+        exchangeNode.getOrderingScheme().ifPresent(orderingScheme -> builder.addAll(orderingScheme.getOrderBy()));
+        Set<Symbol> outputsToRetain = builder.build();
+
+        if (outputsToRetain.size() == exchangeNode.getOutputSymbols().size()) {
+            return Optional.empty();
+        }
+
+        ImmutableList.Builder<Symbol> newOutputs = ImmutableList.builder();
+        List<List<Symbol>> newInputs = new ArrayList<>(exchangeNode.getInputs().size());
+        for (int i = 0; i < exchangeNode.getInputs().size(); i++) {
+            newInputs.add(new ArrayList<>());
+        }
+
+        // Retain used symbols from output list and corresponding symbols from all input lists
+        for (int i = 0; i < exchangeNode.getOutputSymbols().size(); i++) {
+            Symbol output = exchangeNode.getOutputSymbols().get(i);
+            if (outputsToRetain.contains(output)) {
+                newOutputs.add(output);
+                for (int source = 0; source < exchangeNode.getInputs().size(); source++) {
+                    newInputs.get(source).add(exchangeNode.getInputs().get(source).get(i));
+                }
+            }
+        }
+
+        // newOutputs contains all partition, sort and hash symbols so simply swap the output layout
+        PartitioningScheme newPartitioningScheme = new PartitioningScheme(
+                exchangeNode.getPartitioningScheme().getPartitioning(),
+                newOutputs.build(),
+                exchangeNode.getPartitioningScheme().getHashColumn(),
+                exchangeNode.getPartitioningScheme().isReplicateNullsAndAny(),
+                exchangeNode.getPartitioningScheme().getBucketToPartition());
+
+        return Optional.of(new ExchangeNode(
+                exchangeNode.getId(),
+                exchangeNode.getType(),
+                exchangeNode.getScope(),
+                newPartitioningScheme,
+                exchangeNode.getSources(),
+                newInputs.stream().map(ImmutableList::copyOf).collect(toImmutableList()),
```

Comment:
No need to copy the lists here. It's up to ExchangeNode to do a defensive copy if it cares about it.